src/OrchardCore.Modules/OrchardCore.Users.TimeZone/Services/UserTimeZoneService.cs

```diff
@@ -0,0 +1,89 @@
+using OrchardCore.Users.Models;
+using OrchardCore.Users.TimeZone.Models;
+using YesSql;
+
+namespace OrchardCore.Users.TimeZone.Services
+{
+    public class UserTimeZoneService : IUserTimeZoneService
+    {
+        private const string CacheKey = "UserTimeZone";
+
+        private readonly IClock _clock;
+        private readonly IMemoryCache _memoryCache;
+        private readonly IServiceProvider _serviceProvider;
+        private readonly IHttpContextAccessor _httpContextAccessor;
+
+        public UserTimeZoneService(
+            IClock clock,
+            IMemoryCache memoryCache,
+            IServiceProvider serviceProvider,
+            IHttpContextAccessor httpContextAccessor
+            )
+        {
+            _clock = clock;
+            _memoryCache = memoryCache;
+            _serviceProvider = serviceProvider;
+            _httpContextAccessor = httpContextAccessor;
+        }
+
+        public async Task<ITimeZone> GetUserTimeZoneAsync()
+        {
+            string currentTimeZoneId = await GetCurrentUserTimeZoneIdAsync();
+            if (String.IsNullOrEmpty(currentTimeZoneId))
+            {
+                return null;
+            }
+
+            return _clock.GetTimeZone(currentTimeZoneId);
+        }
+
+        public async Task UpdateUserTimeZoneAsync(UserProfile profile)
+        {
+            if (!String.IsNullOrEmpty(_httpContextAccessor.HttpContext.User.Identity.Name))
+            {
+                var session = GetSession();
+
+                var user = await session.Query<User, UserIndex>().Where(x => x.NormalizedUserName == _httpContextAccessor.HttpContext.User.Identity.Name.ToUpper()).FirstOrDefaultAsync();
+                if (user.Properties["UserProfile"] != null && user.Properties["UserProfile"]["TimeZone"] != null)
+                {
+                    user.Properties["UserProfile"]["TimeZone"] = profile.TimeZone;
+                }
+
+                _memoryCache.Set(CacheKey, (string)profile.TimeZone);
+                session.Save(user);
+            }
+        }
+
+        public async Task<string> GetCurrentUserTimeZoneIdAsync()
+        {
+            if (!String.IsNullOrEmpty(_httpContextAccessor.HttpContext.User.Identity.Name))
+            {
+                if (!_memoryCache.TryGetValue(CacheKey, out string timeZoneId))
+                {
+                    var session = GetSession();
+
+                    var user = await session.Query<User, UserIndex>().Where(x => x.NormalizedUserName == _httpContextAccessor.HttpContext.User.Identity.Name.ToUpper()).FirstOrDefaultAsync();
+                    timeZoneId = (string)user.Properties["UserProfile"]["TimeZone"] ?? _clock.GetSystemTimeZone().TimeZoneId;
+
+                    _memoryCache.Set(CacheKey, (string)user.Properties["UserProfile"]["TimeZone"]);
+                }
+
+                return timeZoneId;
+            }
+            else return null;
+        }
+
+        private YesSql.ISession GetSession()
```

Comment:
Just inject `ISession`

Reply:
Tried and it makes saving the data take forever because we create 2 different YesSQL sessions "I think".